spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-page-views.spec.js

```diff
@@ -310,4 +312,43 @@ describe('Google Tag Manager page view tracking', function () {
 
     expect(window.dataLayer[0]).toEqual(expected)
   })
+
+  describe('correctly sets the referrer and virtual_page_view properties', function () {
+    it('when not passed a parameter', function () {
+      GOVUK.analyticsGa4.analyticsModules.PageViewTracker.init()
+
+      expect(window.dataLayer[0]).toEqual(expected)
+    })
+
+    it('when passed a parameter for the referrer', function () {
+      expected.page_view.referrer = 'https://gov.uk/referrer'
+      GOVUK.analyticsGa4.analyticsModules.PageViewTracker.init({
+        referrer: 'https://gov.uk/referrer',
+        dynamic: null
+      })
+
+      expect(window.dataLayer[0]).toEqual(expected)
+    })
+
+    it('when passed a parameter for the virtual_page_view', function () {
+      expected.page_view.dynamic = 'true'
+      GOVUK.analyticsGa4.analyticsModules.PageViewTracker.init({
+        referrer: null,
+        dynamic: 'true'
+      })
+
+      expect(window.dataLayer[0]).toEqual(expected)
+    })
+
+    it('when passed a parameter for the virtual_page_view and referrer', function () {
+      expected.page_view.referrer = 'https://gov.uk/referrer'
+      expected.page_view.dynamic = 'true'
+      GOVUK.analyticsGa4.analyticsModules.PageViewTracker.init({
+        referrer: 'https://gov.uk/referrer',
+        dynamic: 'true'
```

Comment:
You've covered all the possibilities of these two keys existing/not existing very thoroughly here. I'm just curious though - I thought that they're intrinsically linked i.e. if there's a value for `referrer` then `dynamic` is always true, otherwise both are missing. Can you explain the three scenarios being tested here?

Reply:
I think this is sort of linked to the first comment - if we're happy with keeping the code as it is, this means that the `referrer` and `dynamic` properties are being conditionally set by different values on the `dynamicPageViewData` object and I'd just want to be absolutely sure that everything is working as it should in all possible scenarios

Reply:
Been wrestling with this over lunch. I think I'm coming around to the thought that this doesn't feel quite right - we're passing an object that contains two values, but only really using one of them (referrer). I'm quite happy to pass an object (it's more future proof, although I'm not aware of anything else we'd need to pass) but the code and tests currently imply that different values might be passed in that object, when in reality there are only two scenarios: nothing is passed, or referrer and dynamic (set to true) are passed.

Reply:
Cool, thanks for this @andysellick. What I've done is I've changed the implementation slightly so that it's only using a single string parameter now rather than an object. Like you say, I'm not aware of anything further that would need to be passed and, thinking about it, I think passing an object when we only need a single value might not be the most intuitive. If this needs extending in the future we can make it an object but, in the interests of keeping the code more readable in its current form, perhaps just passing the string is best.

Reply:
Thanks @JamesCGDS, this feels much tidier 👍